### PR TITLE
Abs filter

### PR DESCRIPTION
**Pull request by fw42**
The `abs` filter returns the absolute value of a number (i.e. the number itself if it's non-negative and -1 times the number if it's negative).

@pushrax @dylanahsmith @admhlt @freakdesign

### CODE REVIEW

**Comment by fw42** [comment]
Updated to move the BigDecimal to float conversion to the very last step (right before rendering), so that we don't lose precision if `abs` is chained with other filters.

**Comment by fw42** [comment]
Hm actually this might still be losing precision (because `Context#find_variable` calls `to_liquid`). Where's the right place to put this conversion?

**Comment by pushrax** [comment]
Would it break things if we don't convert to float?

**Comment by fw42** [comment]
yea it would.. stringifying big decimals behaves differently, since it uses exponent syntax in some cases
```
irb(main):002:0> 14.42.to_s
=> "14.42"
irb(main):003:0> BigDecimal.new("17.42").to_s
=> "0.1742E2"
```

**Comment by pushrax** [comment]
hmm and any filter can theoretically `to_s` right now, so we have to convert to float in every filter

File: lib/liquid/extensions.rb [hunk 1]
```diff
@@ -25,6 +26,12 @@ def to_liquid
   end
 end
 
+class BigDecimal
+  def to_liquid
+    to_f
```
Comment (dylanahsmith) [inline lib/liquid/extensions.rb]: This basically means that we shouldn't store big decimals in liquid.  I don't think we should do that, and it doesn't seem related to adding a filter.

File: lib/liquid/standardfilters.rb [hunk 2]
```diff
@@ -349,8 +352,7 @@ def default(input, default_value = ''.freeze)
     private
 
     def apply_operation(input, operand, operation)
-      result = Utils.to_number(input).send(operation, Utils.to_number(operand))
-      result.is_a?(BigDecimal) ? result.to_f : result
+      Utils.to_number(input).send(operation, Utils.to_number(operand))
```
Comment (dylanahsmith) [inline lib/liquid/standardfilters.rb]: Shouldn't `public_send` work here?

File: lib/liquid/extensions.rb (hunk 1 above)
Comment (fw42) [inline lib/liquid/extensions.rb]: what do you mean by "store"? The issue is that when converting strings to numbers (and they have a decimal point), we internally use BigDecimal. But if you chain filters together, this is kinda pointless... but I agree that this could be a separate PR. Would you prefer that I just do this convert in the `abs` filter? (like the other number filters do)

Comment (dylanahsmith) [inline lib/liquid/extensions.rb]: Well, if you chain together filters, then we call .to_liquid on the result to avoid using non-liquid values.  The same is true if you assign the result to a variable then retrieve it from a variable tag.  That makes a BigDecimal seem even more pointless.

Comment (fw42) [inline lib/liquid/extensions.rb]: what's your suggestion?

Comment (dylanahsmith) [inline lib/liquid/extensions.rb]: I guess I don't really see the problem with `Utils.to_number(input).abs` since that keeps it as a BigDecimal.

There are some sketch conversions like one from BigDecimal to a Float in the round filter and Utils.to_number converting a Float to BigDecimal.  I'm not sure what is the best way to handle that in a backwards manner though, since the default behaviour for BigDecimal#to_s is weird.

Comment (fw42) [inline lib/liquid/extensions.rb]: yea exactly, the `to_s` thing is the reason why I didn't want to just do `Utils.to_number(input).abs`

Comment (dylanahsmith) [inline lib/liquid/extensions.rb]: Well the same problem would be present if `to_s` was called on the input.  E.g. 

```ruby
irb> Liquid::Template.parse('{{ decimal }}').render("decimal" => BigDecimal.new("12.34"))
=> "0.1234E2"
```

so having the filter convert to a float just makes things less consistent.  I suppose this change makes things more consistent, but it doesn't help for preserving precision.

Comment (fw42) [inline lib/liquid/extensions.rb]: > having the filter convert to a float just makes things less consistent. 

why is that? converting it to float will keep the same to_s behaviour as the other filters have (and all the other number filters also convert to float)

Comment (dylanahsmith) [inline lib/liquid/extensions.rb]: I mean the way `{{ decimal }}` and `{{ decimal | abs }}` are rendered isn't consistent.  I suppose it would make the abs filter consistent with other math filters.

**Comment by fw42** [comment]
updated.. @dylanahsmith do you like this better now?

**Comment by dylanahsmith** [comment]
:+1:
